modules/javafx.graphics/src/main/java/javafx/scene/Node.java

```diff
@@ -8168,29 +8205,107 @@ public String getName() {
      * @see #requestFocus()
      * @defaultValue false
      */
-    private FocusedProperty focused;
+    private final FocusPropertyBase focused = new FocusPropertyBase() {
+        @Override
+        protected PseudoClass getPseudoClass() {
+            return FOCUSED_PSEUDOCLASS_STATE;
+        }
 
-    protected final void setFocused(boolean value) {
-        FocusedProperty fp = focusedPropertyImpl();
-        if (fp.value != value) {
-            fp.store(value);
-            fp.notifyListeners();
+        @Override
+        public String getName() {
+            return "focused";
+        }
+
+        @Override
+        protected boolean notifyListeners() {
+            if (super.notifyListeners()) {
+                notifyAccessibleAttributeChanged(AccessibleAttribute.FOCUSED);
+                return true;
+            }
+            return false;
         }
+
+        @Override
+        public void set(boolean value) {
+            if (get() != value) {
+                super.set(value);
+
+                Node node = Node.this;
+                do {
+                    node.focusWithin.set(value);
+                    node = node.getParent();
+                } while (node != null);
+            }
+        }
+    };
+
+    protected final void setFocused(boolean value) {
+        setFocusQuietly(value, false);
+        notifyFocusListeners();
     }
 
     public final boolean isFocused() {
-        return focused == null ? false : focused.get();
+        return focused.get();
     }
 
     public final ReadOnlyBooleanProperty focusedProperty() {
-        return focusedPropertyImpl();
+        return focused;
     }
 
-    private FocusedProperty focusedPropertyImpl() {
-        if (focused == null) {
-            focused = new FocusedProperty();
+    /**
+     * Indicates whether this {@code Node} should visibly indicate focus.
+     * This flag is set when the node acquires input focus via keyboard navigation,
+     * and it is cleared when the node loses focus, or when {@link #requestFocus()}
+     * is called.
+     *
+     * @defaultValue false
+     * @since 18
+     */
+    private final FocusPropertyBase focusVisible = new FocusPropertyBase() {
+        @Override
+        protected PseudoClass getPseudoClass() {
+            return FOCUS_VISIBLE_PSEUDOCLASS_STATE;
         }
-        return focused;
+
+        @Override
+        public String getName() {
+            return "focusVisible";
+        }
+    };
+
+    public final boolean isFocusVisible() {
+        return focusVisible.get();
+    }
+
+    public final ReadOnlyBooleanProperty focusVisibleProperty() {
+        return focusVisible;
+    }
+
+    /**
+     * Indicates whether this {@code Node} or any of its children currently
```

Comment:
children --> descendants

It isn't just the immediate children that will cause a parent node to report `focusWithin`.